### PR TITLE
Feat/Add-option-load_best-to-TorchForecastingModel

### LINKED ISSUE #2902: [FEATURE] Add option `load_best` to `TorchForecastingModel.fit()` and `fit_from_dataset()`
Related to #2064.

### Summary

We should add an option to our TorchForecastinModel that allows to automatically load (and use) the best model on the validation set at the end of the training process. 

This should only be effective if `val_series` and `val_*_covaraites` are provided, and `save_checkpoints=True` at model creation.

### Potential solution
We could add a param `load_best` (?) to `fit()` and `fit_from_dataset()` to control this.

### Alternative solution
We could instead also add a new param to `__init__()` but I feel it's more appropriate in the `fit*()` methods. Happy to discuss.

### PR DESCRIPTION
**Pull request by cnhwl**
Checklist before merging this PR:
- [x] Mentioned all issues that this PR fixes or addresses.
- [x] Summarized the updates of this PR under **Summary**.
- [x] Added an entry under **Unreleased** in the [Changelog](../CHANGELOG.md).

Fixes #2902.

### Summary
Add a param `load_best` to `TorchForecastinModel` fit() and fit_from_dataset()

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -15,6 +15,7 @@ but cannot always guarantee backwards compatibility. Changes that may **break co
...
 - Added `add_regressor_configs` parameter to the `Prophet` model, enabling component-specific control over `prior_scale`, `mode`, and `standardize` for the future covariates. [#2882](https://github.com/unit8co/darts/issues/2882) by [Ramsay Davis](https://github.com/RamsayDavisWL).
 - 🔴 Increased the decimal places for quantile component names from 2 to 3 for more precise quantiles. (e.g. `component_name_q0.500` for quantile 0.5). This affects quantile forecasts as well as quantiles computed with `TimeSeries.quantile()`. [#2887](https://github.com/unit8co/darts/pull/2786) by [He Weilin](https://github.com/cnhwl).
+- Add an option param `load_best` to `TorchForecastinModel` that allows to automatically load (and use) the best model on the validation set at the end of the training process. [#2903](https://github.com/unit8co/darts/pull/2903) by [He Weilin](https://github.com/cnhwl).
```
Comment (Kurokabe) [inline CHANGELOG.md]: Small typo: TorchForecastin**g**Model (missing g in Forecasting)
```suggestion
- Add an option param `load_best` to `TorchForecastingModel` that allows to automatically load (and use) the best model on the validation set at the end of the training process. [#2903](https://github.com/unit8co/darts/pull/2903) by [He Weilin](https://github.com/cnhwl).
```

**Review by Kurokabe — CHANGES_REQUESTED** [review]
Thanks @cnhwl for the PR. It seems to me that it solves the issue in a neat way.

However, according to the Codecov Report, your main changes (inside `def _train`) are missing coverage. Would it be possible to add tests for that as well?

For instance inside `tests/models/forecasting/test_torch_forecasting_model.py` you might have a test similar to this one (need to be checked if this works):

```python
def test_load_best(self):
        model = RNNModel(...)

        model.fit(self.series, load_best=True)
        best_model_epochs_trained = model.epochs_trained

        model.fit(self.series, load_best=False)
        last_model_epochs_trained  = model.epochs_trained

        assert last_model_epochs_trained != best_model_epochs_trained
```

File: darts/models/forecasting/torch_forecasting_model.py [hunk 2]
```diff
@@ -902,7 +903,10 @@ def fit(
             The number of time steps between consecutive samples, applied starting from the end of the series. The same
             stride will be applied to both the training and evaluation set (if supplied). This should be used with
             caution as it might introduce bias in the forecasts.
-
+        load_best
+            If set to `True`, the model will load the best checkpoint found during training according to the validation
+            loss. This requires `save_checkpoints` to be `True` in the model constructor and a validation set to be
+            provided. Default: ``False``.
```
Comment (dennisbader) [inline darts/models/forecasting/torch_forecasting_model.py]:
```suggestion
            Whether the model should automatically load the best checkpoint found during training according to the
            validation loss. Only effective when `save_checkpoints` was set to `True` in the model constructor and a 
            validation set is passed to the current fit method. Otherwise, it will be ignored. Default: ``False``.
```

File: darts/models/forecasting/torch_forecasting_model.py [hunk 3]
```diff
@@ -1114,21 +1123,31 @@ def fit_from_dataset(
             <https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader>`_.
             By default, Darts configures parameters ("batch_size", "shuffle", "drop_last", "collate_fn", "pin_memory")
             for seamless forecasting. Changing them should be done with care to avoid unexpected behavior.
+        load_best
+            If set to `True`, the model will load the best checkpoint found during training according to the validation
+            loss. This requires `save_checkpoints` to be `True` in the model constructor and `val_dataset` to be
+            provided. Default: ``False``.
```
Comment (dennisbader) [inline darts/models/forecasting/torch_forecasting_model.py]:
```suggestion
            Whether the model should automatically load the best checkpoint found during training according to the
            validation loss. Only effective when `save_checkpoints` was set to `True` in the model constructor and a 
            validation set is passed to the current fit method. Otherwise, it will be ignored. Default: ``False``.
```

File: darts/models/forecasting/torch_forecasting_model.py [hunk 4]
```diff
@@ -1322,6 +1344,30 @@ def _train(
         else:
             trainer.strategy.connect(model)
         self.model = model
+
```
Comment (dennisbader) [inline darts/models/forecasting/torch_forecasting_model.py]: 👍 looks nice :) 

I would suggest we do some adaptions here:
- first (before fitting), we need to check whether checkpointing is activated in general and whether a val set is provided. We need to check proper checkpointing for two cases:
  - Darts automatic checkpointing (via `save_checkpoints=True`)
  - Custom trainer checkpointing (when user provides a custom `trainer` to the fit methods)
  - We can validate these cases by doing some sanity checks (shown below)
- we should skip loading the best if the model doesn't require training
- we need to add tests for full coverage here

Here is my suggestion which should cover these cases (I couldn't yet test whether it works for a custom `trainer`, could you add a test for that):

```py
        self._fit_called = True

        # if model was loaded from checkpoint (when `load_ckpt_path is not None`) and model.fit() is called,
        # we resume training
        ckpt_path = self.load_ckpt_path
        self.load_ckpt_path = None

        if load_best:
            ckpt_callback: Optional[ModelCheckpoint] = trainer.checkpoint_callback
            ckpt_activated = ckpt_callback is not None and hasattr(ckpt_callback, "best_model_path")
            if not ckpt_activated or val_loader is None:
                logger.warning(
                    "Loading the best model will be skipped (`load_best` is ignored), as it requires "
                    "active checkpointing and a validation set to be provided to the current fit method."
                    "If not using a custom `trainer`, make sure to set `save_checkpoints=True` at model creation. "
                    "Otherwise, make sure the custom `trainer` uses a pytorch-lightning `ModelCheckpoint` callback."
                )
                load_best = False
        else:
            ckpt_callback = None

        if self._requires_training:
            trainer.fit(
                model,
                train_dataloaders=train_loader,
                val_dataloaders=val_loader,
                ckpt_path=ckpt_path,
            )
            if load_best:
                best_model_path = ckpt_callback.best_model_path
                logger.info(
                    f"Loading best model from checkpoint: '{os.path.basename(best_model_path)}'"
                )
                model = self._load_from_checkpoint(best_model_path)
        else:
            trainer.strategy.connect(model)
        self.model = model
        self.trainer = trainer
```

File: darts/tests/models/forecasting/test_torch_forecasting_model.py [hunk 5]
```diff
@@ -2515,6 +2515,43 @@ def test_predict_after_fit_from_dataset(self):
         assert pred3.time_index.equals(pred1.time_index)
         assert pred3.static_covariates.equals(series[0].static_covariates)
 
+    def test_load_best(self, tmpdir_fn):
```
Comment (dennisbader) [inline darts/tests/models/forecasting/test_torch_forecasting_model.py]: great test :)

File: darts/tests/models/forecasting/test_torch_forecasting_model.py [hunk 6]
```diff
@@ -2515,6 +2515,43 @@ def test_predict_after_fit_from_dataset(self):
...
+
+        # The model trained to the end should have n_epochs trained
+        assert last_model_epochs_trained == n_epochs
+        # The model loading the best checkpoint should have trained fewer epochs (or equal if best is last)
+        # With our crafted data, it's highly likely to be less.
+        assert best_model_epochs_trained < last_model_epochs_trained
+        # We can also check that the trainer object confirms the best model was loaded
+        best_model_path = model_best.trainer.checkpoint_callback.best_model_path
+        assert best_model_path and os.path.exists(best_model_path)
+
```
Comment (dennisbader) [inline darts/tests/models/forecasting/test_torch_forecasting_model.py:2597]: let's check prediction as well
```suggestion
        # predictions must work and be different
        preds_last = model_last.predict(n=1)
        preds_best = model_best.predict(n=1)
        assert preds_last != preds_best

```

File: darts/tests/models/forecasting/test_torch_forecasting_model.py (hunk 5 above)
Comment (dennisbader) [inline darts/tests/models/forecasting/test_torch_forecasting_model.py]: Can you parametrize this test to check the two cases (using `@pytest.mark.parametrize`):
- Case A: Darts' automatic checkpointing: what you currently have
- Case B: User provides a custom `trainer` to `fit()` with a ModelCheckpoint callback

File: darts/tests/models/forecasting/test_torch_forecasting_model.py
Comment (dennisbader) [inline darts/tests/models/forecasting/test_torch_forecasting_model.py:1]: Could you add a test for the case that where `load_best` is ignored? E.g. user specifies `load_best=True`, but `save_checkpoints=False` or no validation set was passed?

**Review by dennisbader — CHANGES_REQUESTED** [review]
Thanks a lot for another great PR @cnhwl 🚀 

I really like your approach :) 

I added a couple of suggestions to cover more edge cases. Let me know what you think.

File: CHANGELOG.md [hunk 7]
```diff
@@ -16,6 +16,7 @@ but cannot always guarantee backwards compatibility. Changes that may **break co
...
 - Added `add_regressor_configs` parameter to the `Prophet` model, enabling component-specific control over `prior_scale`, `mode`, and `standardize` for the future covariates. [#2882](https://github.com/unit8co/darts/issues/2882) by [Ramsay Davis](https://github.com/RamsayDavisWL).
 - 🔴 Increased the decimal places for quantile component names from 2 to 3 for more precise quantiles. (e.g. `component_name_q0.500` for quantile 0.5). This affects quantile forecasts as well as quantiles computed with `TimeSeries.quantile()`. [#2887](https://github.com/unit8co/darts/pull/2786) by [He Weilin](https://github.com/cnhwl).
+- Add an option param `load_best` to `TorchForecastingModel` that allows to automatically load (and use) the best model on the validation set at the end of the training process. [#2903](https://github.com/unit8co/darts/pull/2903) by [He Weilin](https://github.com/cnhwl).
```
Comment (dennisbader) [inline CHANGELOG.md]:
```suggestion
- Added parameter `load_best` to `TorchForecastingModel.fit()` that allows to automatically load (and use) the best model on the validation set at the end of the training process. [#2903](https://github.com/unit8co/darts/pull/2903) by [He Weilin](https://github.com/cnhwl).
```

File: darts/models/forecasting/torch_forecasting_model.py [hunk 8]
```diff
@@ -1466,7 +1512,25 @@ def lr_find(
...
+        ) = params
+        trainer, model, train_loader, val_loader = self._setup_for_train(
+            train_dataset=train_dataset,
+            val_dataset=val_dataset,
+            trainer=trainer,
+            verbose=verbose,
+            epochs=epochs,
+            dataloader_kwargs=dataloader_kwargs,
+        )
+
```
Comment (dennisbader) [inline darts/models/forecasting/torch_forecasting_model.py]: It looks to me like we could revert these changes?

```suggestion
```

File: darts/tests/models/forecasting/test_torch_forecasting_model.py [hunk 9]
```diff
@@ -2515,6 +2517,156 @@ def test_predict_after_fit_from_dataset(self):
...
+        model_kwargs = {
+            "input_chunk_length": 12,
+            "model": "RNN",
+            "hidden_dim": 10,
+            "n_rnn_layers": 1,
+            "n_epochs": n_epochs,
+            "work_dir": tmpdir_fn,
+            "random_state": 42,
+            "save_checkpoints": save_checkpoints,
+            "model_name": f"test_load_best_ignored_{save_checkpoints}_{val_series_provided}",
```
Comment (dennisbader) [inline darts/tests/models/forecasting/test_torch_forecasting_model.py]: Let's reduce it to what is required for the test

```suggestion
            "n_epochs": n_epochs,
            "work_dir": tmpdir_fn,
            "random_state": 42,
            "save_checkpoints": save_checkpoints,
```

File: darts/tests/models/forecasting/test_torch_forecasting_model.py [hunk 10]
```diff
@@ -2515,6 +2517,156 @@ def test_predict_after_fit_from_dataset(self):
...
+        # The model trained to the end should have n_epochs trained
+        assert last_model_epochs_trained == n_epochs
+        # The model loading the best checkpoint should have trained fewer epochs (or equal if best is last)
+        # With our crafted data, it's highly likely to be less.
+        assert best_model_epochs_trained < last_model_epochs_trained
+
+        # We can also check that the trainer object confirms the best model was loaded
+        trainer_to_check = trainer_best if use_custom_trainer else model_best.trainer
+        best_model_path = trainer_to_check.checkpoint_callback.best_model_path
+        assert best_model_path and os.path.exists(best_model_path)
```
Comment (dennisbader) [inline darts/tests/models/forecasting/test_torch_forecasting_model.py:2596]: Nice :) And now I as a last thing we should check that loading the best model checkpoint from `model_last` gives identical predicitons to `model_best`.

E.g. something like below


```py
model_last_best = RNNModel.load_from_checkpoint(
    model_name=model_last.model_name,
    work_dir=model_last.work_dir,
    best=True,
)

preds_last_best = model_last_best.predict(n=1)
assert preds_last_best == preds_best
```

**Review by dennisbader — CHANGES_REQUESTED** [review]
Thanks for the updates @cnhwl 🚀 

I added some last minor suggestion but then it's ready to be merged :)

File: darts/models/forecasting/torch_forecasting_model.py (hunk 8 above)
Comment (cnhwl) [reply darts/models/forecasting/torch_forecasting_model.py]: I attempted to remove the code here, but `model`, `train_loader` and `val_loader` would become Undefined, so I have added the code back.

**Review by dennisbader — APPROVED** [review]
Awesome @cnhwl 🚀 Great PR and thanks a lot 🥳 

I pushed some last minor changes. I'll merge this one after the tests have passed 💯
